discovery_engine_core/core/src/storage/sqlite.rs

```diff
@@ -15,182 +15,448 @@
+        }
+
+        // The amount of documents that we can store via bulk inserts
+        // (https://docs.rs/sqlx-core/latest/sqlx_core/query_builder/struct.QueryBuilder.html#method.push_values)
+        // is limited by the sqlite bind limit.
+        // bind_limit divided by the number of fields in the largest tuple (NewsResource)
+        let documents = documents.iter().take(BIND_LIMIT / 9);
+
+        // Begin transaction
+        let mut tx = self
+            .pool
+            .begin()
+            .await
+            .map_err(|err| storage::Error::Database(err.into()))?;
+
+        // insert id into Document table (fk of HistoricDocument)
+        let mut query_builder = QueryBuilder::new("INSERT INTO Document (id) ");
+        query_builder.push_values(documents.clone(), |mut stm, doc| {
+            stm.push_bind(doc.id.as_uuid());
+        });
+        query_builder
+            .build()
+            .persistent(false)
+            .execute(&mut tx)
+            .await
+            .map_err(|err| storage::Error::Database(err.into()))?;
+
+        // insert id into HistoricDocument table
+        let mut query_builder = QueryBuilder::new("INSERT INTO HistoricDocument (documentId) ");
+        query_builder.push_values(documents.clone(), |mut stm, doc| {
+            stm.push_bind(doc.id.as_uuid());
+        });
+        query_builder
+            .build()
+            .persistent(false)
+            .execute(&mut tx)
+            .await
+            .map_err(|err| storage::Error::Database(err.into()))?;
+
+        // insert data into NewsResource table
+        let mut query_builder = QueryBuilder::new("INSERT INTO NewsResource (documentId, title, snippet, topic, url, image, datePublished, source, market) ");
+        query_builder.push_values(documents.clone(), |mut stm, doc| {
+            stm.push_bind(doc.id.as_uuid())
+                .push_bind(&doc.resource.title)
+                .push_bind(&doc.resource.snippet)
+                .push_bind(&doc.resource.topic)
+                .push_bind(doc.resource.url.to_string())
+                .push_bind(doc.resource.image.as_ref().map(ToString::to_string))
+                .push_bind(&doc.resource.date_published)
+                .push_bind(&doc.resource.source_domain)
+                .push_bind(format!("{}{}", doc.resource.language, doc.resource.country));
+        });
+        query_builder
+            .build()
+            .persistent(false)
+            .execute(&mut tx)
+            .await
+            .map_err(|err| storage::Error::Database(err.into()))?;
+
+        // insert data into NewscatcherData table
+        let mut query_builder =
+            QueryBuilder::new("INSERT INTO NewscatcherData (documentId, domainRank, score) ");
+        query_builder.push_values(documents.clone(), |mut stm, doc| {
+            stm.push_bind(doc.id.as_uuid())
+                .push_bind(doc.resource.rank as u32)
+                .push_bind(doc.resource.score.unwrap_or_default());
```

Comment:
Can we use `unwrap_or_default` here? in the NewscatcherData table we define `score` as not null so we need a value or we allow it to be null.

Reply:
no that table should have a nullable value, the score only applies to newscatcher documents from search not from headlines.

Reply:
made score nullable